wal/src/message_queue_impl/region_meta.rs

```diff
@@ -0,0 +1,595 @@
+            entries,
+        })
+    }
+}
+
+/// Region meta data.
+#[derive(Default, Debug)]
+struct RegionMetaInner {
+    table_metas: HashMap<TableId, TableMeta>,
+    /// It will fall behind the high watermark for ensuring successive
+    /// increment.
+    local_latest_offset: AtomicI64,
+}
+
+/// Wrapper for the [TableMetaInner].
+// TODO: will be made use of later.
+#[allow(unused)]
+#[derive(Debug)]
+struct TableMeta {
+    table_id: TableId,
+    /// The race condition may occur between writer thread
+    /// and background flush tread.
+    inner: Mutex<TableMetaInner>,
+}
+
+// TODO: will be made use of later.
+#[allow(unused)]
+impl TableMeta {
+    fn new(table_id: TableId) -> Self {
+        Self {
+            table_id,
+            inner: Mutex::new(TableMetaInner::default()),
+        }
+    }
+
+    async fn prepare_for_write(&self) -> SequenceNumber {
+        let inner = self.inner.lock().await;
+        inner.next_sequence_num
+    }
+
+    async fn update_after_write(&self, write_offset_range: (Offset, Offset), updated_num: u64) {
+        let mut inner = self.inner.lock().await;
+        let old_next_sequence_num = inner.next_sequence_num;
+        inner.next_sequence_num += updated_num;
+
+        // Update the mapping and high water mark.
+        let _ = inner
+            .start_sequence_offset_mapping
+            .insert(old_next_sequence_num, write_offset_range.0);
+        inner.current_high_watermark = write_offset_range.1 + 1;
+    }
+
+    async fn mark_deleted(&self, latest_marked_deleted: SequenceNumber) {
+        let mut inner = self.inner.lock().await;
+        assert!(latest_marked_deleted <= inner.next_sequence_num);
+        inner.latest_marked_deleted = latest_marked_deleted;
+
+        // Update the mapping, keep the range in description.
+        inner
+            .start_sequence_offset_mapping
+            .retain(|k, _| k >= &latest_marked_deleted);
+    }
+
+    async fn get_meta_data(&self) -> Result<TableMetaData> {
+        let inner = self.inner.lock().await;
+
+        // Only two situations exist:
+        // + no log of the table has ever been written
+        //  (next sequence num == latest marked deleted).
+        // + all logs of the table can be deleted
+        //  (next_sequence_num > latest_marked_deleted).
+        if inner.next_sequence_num == inner.latest_marked_deleted {
+            Ok(TableMetaData {
+                table_id: self.table_id,
+                next_sequence_num: inner.next_sequence_num,
+                latest_marked_deleted: inner.latest_marked_deleted,
+                current_high_watermark: inner.current_high_watermark,
+                safe_delete_offset: None,
+            })
+        } else {
+            let offset = inner
+                .start_sequence_offset_mapping
+                .get(&inner.latest_marked_deleted);
+            ensure!(inner.next_sequence_num > inner.latest_marked_deleted && offset.is_some(),
```

Comment:
It will be more clear for troubleshooting if we divide this check into two parts: 
```
ensure!(inner.next_sequence_num > inner.latest_marked_deleted, ...);
let offset = inner..start_sequence_offset_mapping
                .get(&inner.latest_marked_deleted)
                .context()?;
```
```